[LLD] Improve linker script handing in LLD

hongyu-dev opened the pull request:
The current implementation of linker scripts in LLD faces challenges due to inadequate support and tooling, particularly for tasks such as writing and debugging. To address these issues, especially in embedded development, improvements to the lexer and parser in LLD are proposed, including better diagnostics, support for LTO code generation, and potentially creating an intermediate representation for linker scripts within LLD.

One key proposal is to replace the use of lambda functions (`Expr`) with new, well-defined expression types called `ScriptExpr`. Currently, expressions in linker scripts are represented as lambdas, which, while practical, hinder the separation of expression evaluation from the parser and make managing state opaque and challenging. The proposed `ScriptExpr` types aim to make this state systematically accessible through a well-defined context, improving code organization, maintainability, and extensibility.

The benefits of this transition include improved debuggability, better performance through efficient memory allocation, and more flexible and efficient expression handling. The new types—`ConstantExpr`, `FallbackExpr`, `UnaryExpr`, and `BinaryExpr`—are inspired by the `MCExpr` in LLVM and provide a clearer structure for linker script expressions.

For more details, please check our RFC: https://discourse.llvm.org/t/rfc-improve-linker-script-handing-in-lld/80866

hongyu-dev committed:
[ELF] Added ScriptExpr, DynamicExpr, ConstantExpr, BinaryExpr, and
UnaryExpr

The use of `using Expr = std::function<ExprValue()>;` for representing
expressions in linker scritps. This commit plans to replace it by using
syntactic objects since the current way potentially complictaes further
analysis and transformations.

hongyu-dev committed:
[ELF] Updated ScriptExpr subclasses

hongyu-dev committed:
[ELF] Updated function return type from ScriptExpr to ScriptExpr*

Otherwise, we have error because of type slicing

hongyu-dev committed:
[ELF] Replace all Expr to ScriptExpr*

UnaryExpr and BinaryExpr will be added once we can run all with
DynamicExpr successfully. This change contains bugs that do not pass
most LTO tests.

hongyu-dev committed:
[ELF] Fixed nullptr bugs for ScriptExpr* variables

This commit still contains tests failure.

hongyu-dev committed:
[ELF] Binary operation updated

hongyu-dev committed:
[ELF] Updated BinaryExpr symbol assignmemnt

hongyu-dev committed:
[ELF] Removed combine and updated BinaryExpr for evaluate statements

hongyu-dev committed:
[ELF] Updated BinaryExpr

hongyu-dev committed:
[ELF] Simplified nested ScriptExpr with BinaryExpr

hongyu-dev committed:
[ELF] Removed getExprValueAlignValue

hongyu-dev committed:
[ELF] Update llvm_unreachable to error handling

mysterymath commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -131,6 +131,156 @@ uint64_t ExprValue::getSectionOffset() const {
+  switch (kind_) {
+  case ExprKind::Constant:
+    return static_cast<const ConstantExpr *>(this)->getConstantExpr();
+  case ExprKind::Dynamic: {
```
Having second thoughts about the name "Dynamic"; that word implies change or motion, which doesn't actually seem to fit here, since the logic doesn't actually change. It's a variable, not a time series. Maybe "Generic"? "Functor"?

yugier replied:
I changed the name to `FallbackExpr` but we can always change it if we have a better one.

mysterymath commented on lld/ELF/ScriptParser.cpp:
```diff
@@ -1471,21 +1356,22 @@ static bool isValidSymbolName(StringRef s) {
-    Expr e = readPrimary();
-    return [=] { return ~e().getValue(); };
+    ScriptExpr *e = readPrimary();
+    // return [=] { return ~e().getValue(); };
```
This comment seems unnecessary now.

yugier replied:
Fixed!

hongyu-dev committed:
[ELF] Delete getExpr and updated getExprValue for all subclasses of
ScriptExpr

yugier commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -131,6 +131,156 @@ uint64_t ExprValue::getSectionOffset() const {
   return getValue() - getSecAddr();
 }
 
+std::function<ExprValue()> ScriptExpr::getExpr() const {
```
Fixed!

yugier commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,111 @@ struct ExprValue {
+  std::function<ExprValue()> getConstantExpr() const {
+    return [=] { return val_; };
+  }
+  ExprValue getConstantExprValue() const { return ExprValue(val_); }
```
Fixed!

hongyu-dev committed:
[ELF] Replcaed std::unordered_map with llvm StringSwitch in LinkerScript

yugier commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -131,6 +131,156 @@ uint64_t ExprValue::getSectionOffset() const {
+}
+
+BinaryExpr::Op BinaryExpr::stringToOp(const StringRef op) {
+  static const std::unordered_map<llvm::StringRef, Op> opMap = {
```
Fixed!

hongyu-dev committed:
[ELF] Updated some DynamicExpr with ConstantExpr

smithp35 commented on lld/ELF/LinkerScript.h:
```diff
@@ -161,8 +264,12 @@ struct MemoryRegion {
 
-  uint64_t getOrigin() const { return origin().getValue(); }
-  uint64_t getLength() const { return length().getValue(); }
+  uint64_t getOrigin() const {
```
Is it a precondition that origin be non-null if getOrigin() is called? If so it could be worth doing an assert(origin != nullptr) instead of picking a default value of 0.

Same for length below.

At a glance it looks like origin and length are passed in via the constructor and it looks like they can't be nullptr.

yugier replied:
Thank you for the review! I removed the precondition and added assert for `origin` and `length`. Yes I think they should not be `nullptr` since they are passed in the constructor but still have to make sure that they are not passed in as `nullptr` so an assert would be perfect here.

smithp35 reviewed (commented):
Just had a small amount of time to skim through. One possible advantage of an AST is that it should make it possible to pretty print the linker script, or at least parts of it for use in diagnostics. Doing that might influence the design choice for the AST nodes. Just thinking about how a node that contains an arbitrary `std::function` would be printed?

hongyu-dev committed:
[ELF] Added UnaryExpr

mysterymath commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,106 @@ struct ExprValue {
+  ExprValue getExprValue() const;
+
+private:
+  StringRef op_;
```
Does a char work?

yugier replied:
yes it would but I was also thinking if I should keep the same coding pattern like `BinaryExpr` to create `enum class opcode` for `UnaryExpr`.

hongyu-dev committed:
[ELF] Deleted Expr with std::function and changed ScriptExpr to Expr

hongyu-dev committed:
[ELF] Replaced DynamicExpr with FallbackExpr

"Dynamic" implies change or motion which does not fit here. We changed
the name to FallbackExpr since we are using FallbackExpr to carry the
std::function for linker script expressions that cannot be represented
as ConstantExpr, BinaryExpr, and UnaryExpr.

We plan to remove all lambda functions for linker script expression and
FallbackExpr is transition state.

yugier commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,111 @@ struct ExprValue {
 // Later, we evaluate the expression by calling the function.
 using Expr = std::function<ExprValue()>;
 
+class ScriptExpr {
```
Updated `ScriptExpr` to `Expr` and added some comments but not sure if they are good ones.

yugier commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,106 @@ struct ExprValue {
+private:
+  Op opcode;
+  const ScriptExpr *LHS, *RHS;
+  std::string loc_;
```
Not sure about this one either. I saw other places with `loc` information based `std::string` but I can double check here later once we can let all Expr types have location information.

hongyu-dev committed:
[ELF] Removed precondtion check for getOrigin and getLength in
LinkerScript

yugier commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,106 @@ struct ExprValue {
+  uint64_t val_;
+};
+
+class DynamicExpr : public ScriptExpr {
```
I agree that the name `DynamicExpr` is confusing since dynamic has a special meaning in the linker while what we have here has nothing to do with dynamic. I changed the name to `FallbackExpr` since we are temporarily using `FallbackExpr` to carry the lambdas that we cannot represent as `BinaryExpr`, `UnaryExpr`, and `ConstantExpr`. Yes potentially we can add more sub-classes to completely replace all lambda functions but we are still trying to figure  out how to separate information and function calls in the lambda functions; `FallbackExpr` is more like a transition state for eventually removing all lambdas functions we have. 

Hopefully I describe the motivation for `DynamicExpr` or `FallbackExpr` clearly here and please let me know if anything I can elaborate here. Any feedback or suggestions on extracting information we need in lambda functions out so we can remove them would be appreciated!

yugier commented on lld/ELF/LinkerScript.h:
```diff
@@ -74,6 +75,106 @@ struct ExprValue {
+  BinaryExpr(StringRef op, ScriptExpr *LHS, ScriptExpr *RHS,
+             const std::string loc)
+      : ScriptExpr(ExprKind::Binary), LHS(LHS), RHS(RHS), loc_(loc) {
+    opcode = stringToOp(op);
```
Fixed! Thank you!
